### PR TITLE
Allow setting ChannelCacheSize in RabbitMQ binder

### LINKED ISSUE #368: Rabbit Binder ChannelCacheSize Property
As a user, I would like to configure the `channelCacheSize` property on the `CachingConnectionFactory`.

http://stackoverflow.com/questions/35563064/processing-messages-through-namedchannels-with-prefetch-1/35584333#35584333

### PR DESCRIPTION
**Pull request by ilayaperumalg**
- Set ChannelCacheSize property for the binder's CacheConnectionFactory

- Add test

This resolves #368

### CODE REVIEW

File: spring-cloud-stream-binders/spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/config/RabbitMessageChannelBinderConfiguration.java [hunk 1]
```diff
@@ -45,15 +49,19 @@
...
+
 	@Autowired
 	private SpringRabbitMQProperties springRabbitMQProperties;
 
 	@Bean
 	RabbitMessageChannelBinder rabbitMessageChannelBinder() {
+		if (rabbitConnectionFactory instanceof CachingConnectionFactory) {
+			((CachingConnectionFactory) rabbitConnectionFactory).
+					setChannelCacheSize(springRabbitMQProperties.getChannelCacheSize());
+		}
```
Comment (garyrussell) [inline spring-cloud-stream-binders/spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/config/RabbitMessageChannelBinderConfiguration.java]: This just seems wrong to me (reconfiguring Boot's object) - given that we are already overriding the `RabbitProperties` from boot auto-configure, we should really disable the boot auto-config altogether here and create our own factory.

For example, the connection factory `addresses` is populated from `RabbitProperties` but the binder `addresses` is populated from `SpringRabbitMQProperties`. Yes, the source of the property is the same but it's a bit smelly IMHO.

Either `SpringRabbitMQProperties` should just have those properties over and above those in `RabbitProperties` or we should do all the wiring here.

I think I prefer the latter, at least until (or unless) the `channelCachesize` property is added to boot.

I am just not keen on reconfiguring boot's object here.

**Comment by ilayaperumalg** [comment]
+1 for addressing this in `spring-boot` itself as I felt the same when working on this.

**Comment by mbogoevici** [comment]
Can we just close this until Boot supports it then?

**Comment by garyrussell** [comment]
Boot PR: https://github.com/spring-projects/spring-boot/pull/5303

**Comment by ilayaperumalg** [comment]
Given the boot PR https://github.com/spring-projects/spring-boot/pull/5303 is scheduled for **1.4.0.M2**, are we okay waiting for this fix post GA as I believe we can not depend on milestone version of boot?

Re-opening this PR to clarify the question above.

**Comment by mbogoevici** [comment]
@garyrussell can confirm or deny, but this doesn't seem to be a particular issue that Stream should worry about - we don't do request reply scenarios, and we don't open temporary channels all that often. If anything, we can set up a decent default until Boot supports it.

**Comment by garyrussell** [comment]
@mbogoevici No; it has nothing to do with request/reply; the cache is for channels to talk to RabbitMQ over, not queues (similar to a JMS Session).

The problem is on each send a channel is borrowed from the cache; consumer channels are long-lived but producer channels are very short-lived. The default cache size is only 1 so, under extreme conditions, we end up creating/destroying many channels instead of caching them.

**Comment by garyrussell** [comment]
The BOOT PR completely solves it, so it's not really an issue for Stream per se; but I would like to see us clean up the binder config (duplicate properties) before RC, though.

**Comment by mbogoevici** [comment]
Ok, sorry - I was confused by the initial issue originating in an XD job launch scenario. I would close this as is though and keep just https://github.com/spring-cloud/spring-cloud-stream/pull/387 for the binder config.

**Comment by garyrussell** [comment]
Do you mean to augment that PR to remove the duplicate properties?

**Comment by mbogoevici** [comment]
Yes, or create a new one for removing the duplicates specifically. This issue can be closed IMO.

**Pull request closed without merging**